feat/add latency support for trtllm bench

danielafrimi opened the pull request:
@rakib-hasan 
id like to get a review from you :) 

most of the changes are taken from your PR (found only throughput  support).

danielafrimi committed:
add latency support for trtllm bench

Signed-off-by: Daniel Afrimi <dafrimi@nvidia.com>

refactor

Signed-off-by: Daniel Afrimi <dafrimi@nvidia.com>

fix review

Signed-off-by: Ubuntu <dafrimi@nvidia.com>

refactor review

Signed-off-by: Ubuntu <dafrimi@nvidia.com>

wip

Signed-off-by: Ubuntu <dafrimi@nvidia.com>

wip

Signed-off-by: Ubuntu <dafrimi@nvidia.com>

refactor

Signed-off-by: Ubuntu <dafrimi@nvidia.com>

refactor

Signed-off-by: Ubuntu <dafrimi@nvidia.com>

rebase

Signed-off-by: Ubuntu <dafrimi@nvidia.com>

danielafrimi commented:
/bot run

tensorrt-cicd commented:
[PR_Github #2941](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2941/) [ run ] triggered by Bot

danielafrimi commented on tensorrt_llm/bench/benchmark/low_latency.py:
```diff
@@ -134,21 +178,71 @@ def latency_command(
     iteration_writer = IterationWriter(iteration_log)
 
+    # Initialize the HF tokenizer for the specified model.
+    # ignore_eos = True if runtime_config.decoding_config.decoding_mode == SpeculativeDecodingMode.NONE else False # TODO (dafrimi): nto sure where to locate this line since it requires the runConfig, but tokenizer is used to get the dataset
```
@rakib-hasan WDYT? not sure where to locate it

Naveassaf replied:
It **looks like** `ignore_eos` is only used for getting the `eos_id` and `pad_id`. 

Meanwhile, `eos_id` and `pad_id` are only used for creating the `SamplingParameters`. So why not move lines 198, 200 and 201 to be right before the creating of the `SamplingParams` (line ~290)? Then you could have:
```
ignore_eos = True if runtime_config.decoding_config.decoding_mode == SpeculativeDecodingMode.NONE else False
```
since youve already created the runtime config

rakib-hasan replied:
IIUC, this reordering is all because of `runtime_config.decoding_config.decoding_mode`, right?
In that case, can we have an extra API just to get that early? That way `runtime_config` can be created once all the extra options are populated.
For the TRT case, it will come from the engine config (we need to read the config twice, which should be fine for now)
For the PyTorch case, it is currently empty for the throughput benchmark (possibly since SD is not used for throughput usecase) and since PyTorch doesn't support the older SD techniques yet. In that case, we can just keep that empty for now like the throughput case and say that it is a limitation until the PyTorch path supports SD techniques.
Does that sound alright?

rakib-hasan replied:
Just to confirm, with that extra API, this reorder shouldn't be needed, right?

danielafrimi replied:
look on the comment i added (line 288). so we wont need extra API for this case. we still reorder the `exec_settings`, since we fetch need to fetch the setting in the pytorch flow which uses the metadata (from `create_dataset_from_stream`). WDYT?

danielafrimi commented:
/bot run

tensorrt-cicd commented:
[PR_Github #2943](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2943/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #2941](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2941/) [ run ] completed with state `ABORTED`

Naveassaf commented on tensorrt_llm/bench/benchmark/low_latency.py:
```diff
@@ -134,21 +178,71 @@ def latency_command(
+
+    # Engine configuration parsing for PyTorch backend
+    kwargs = {}
+    if backend and backend.lower() in ["pytorch", "autodeploy"]:
```
Id consider factoring this out to some const up top:
```
SUPPORTED_BACKENDS = ["pytorch", "autodeploy", "cpp"]
```
having this as the `--backend`'s choices, and have the default be `cpp`. IMO its confusing to have the default not be an option.

rakib-hasan replied:
Agree on the refactor. Not just up top, but in some utils so that both `low_latency.py` and `throughput.py` can use the same.
However, for default, I would go with pytorch.

FrankD412 replied:
Pytorch is the correct default -- it just got changed [here](https://github.com/NVIDIA/TensorRT-LLM/pull/5312) recently.

Naveassaf commented on tensorrt_llm/bench/benchmark/low_latency.py:
```diff
@@ -134,21 +178,71 @@ def latency_command(
     medusa_choices = params.pop("medusa_choices")
 
-    # Reporting Options
+    # # Reporting Options
```
extra `#`

Naveassaf commented on tensorrt_llm/bench/benchmark/low_latency.py:
```diff
@@ -134,21 +178,71 @@ def latency_command(
+                "support this dataset.")
+
+    exec_settings["model"] = model
+    engine_tokens = exec_settings["settings_config"]["max_num_tokens"]
```
Are you sure this is part of the `settings_config` for the PyTorch flow? I was under the impression its not currently supported at the attention backend level, but I definitely might be wrong.

rakib-hasan replied:
AFAIK, `settings_config` is valid for PyTorch flow.

Naveassaf reviewed (commented):
Nice change :)

tensorrt-cicd commented:
[PR_Github #2943](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2943/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #2064](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/2064/) completed with status: 'FAILURE'

rakib-hasan commented:
@danielafrimi  The other issue you mentioned, making the tokenizer be optional for multimodal: can you add that change to both low_latency and throughput case? That should be a simple change to add. If it turns out to be complicated, then you can certainly create a new PR.

danielafrimi commented:
@rakib-hasan about the issue which makes the tokenizer be optional for multimodal, its can only be optional ion the first phase when we prepare the dataset - can add that there in the PR. 

for running the `trtllm-bench` for the NVILA i initialled the NVILA tokenizer as they did in modeling_vila.py (will create a different PR)

danielafrimi commented on tensorrt_llm/bench/benchmark/low_latency.py:
```diff
@@ -179,45 +270,41 @@ def latency_command(
 
-    llm = None
-    kwargs = runtime_config.get_llm_args()
+        ignore_eos = True if runtime_config.decoding_config.decoding_mode == SpeculativeDecodingMode.NONE else False
```
@rakib-hasan what to you think on this change? we needed the runConfig only for getting the eos/pad id to sampling params. in this change we already have the runtime_config for getting this so TRT + pytorch will work as well.

rakib-hasan replied:
LGTM

FrankD412 replied:
So a caveat here that was brought to my attention yesterday. Models can have more than one EOS token. For example, Llama 4 has three different EOS tokens which are used in different scenarios and can affect the behavior of speculative decoding. Are we sure we're going to get the token we want from simply asking for the tokenizer's EOS?

danielafrimi commented:
/bot run

tensorrt-cicd commented:
[PR_Github #3055](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/3055/) [ run ] triggered by Bot

symphonylyh commented on tensorrt_llm/bench/benchmark/low_latency.py:
@FrankD412 there is an offline discussion on whether this low-latency trtllm-bench is still needed. because for DS and Llama4 we just run regular throughput benchmark by setting different concurrency, so small concurrency is naturally low latency scenarios?
it's still relevant for the TRT workflow by turning on low-latency plugins, but for pytorch workflow it's irrelevant, right?

FrankD412 replied:
@symphonylyh -- yeah, this subcommand has become legacy at this point for the TRT workflow.

FrankD412 replied:
Also, for context -- I've been trying to get around to clean up the subcommands. I have an idea for it. Maybe we should have an offline discussion about what I'm thinking? @rakib-hasan @symphonylyh @danielafrimi

danielafrimi commented on tensorrt_llm/bench/benchmark/utils/general.py:
```diff
@@ -22,6 +22,9 @@
     QuantAlgo.NVFP4.value: "fp8",
 }
 
+SUPPORTED_BACKENDS = ["pytorch", "autodeploy", "cpp"]
+PYTHON_SUPPORTED_BACKENDS = ["pytorch", "autodeploy"]
```
done

danielafrimi commented:
@rakib-hasan Can we merge it?

rakib-hasan commented:
Hi @danielafrimi , as Frank mentioned that `latency` side of the benchmark is just legacy at this point, I will let him decide. 

@FrankD412 do you want this PR to be merged or closed?

If merging, @danielafrimi we need to resync and rerun the CI as it has been 2 weeks since the last run.

rakib-hasan commented:
Sounds good. 

I am not sure whether @danielafrimi is planning to use the latency path or it was just for feature parity with the throughput path. If it is the latter, then this PR should be closed and maybe a new PR for deprecating the latency path would reduce future confusions.

rakib-hasan commented:
Hey @danielafrimi . Apologies. I wrote a response but forgot to send. 
For latency numbers, you can enable the streaming (`--streaming`) in the throughput command. Would that be sufficient for your use-case? 

If so, we can close this one as `latency` path won't be needed. 

If not, we can certainly merge this and note what is missing so that we can add that as part of the refactor that Frank is going to work on.

FrankD412 commented:
@rakib-hasan -- just following up on this one. What's the current status looking like?

rakib-hasan commented:
> @rakib-hasan -- just following up on this one. What's the current status looking like?

I don't know. Not sure if @danielafrimi is OOO or not.

danielafrimi commented:
Hi, sorry for the late response i was OOO. 
didnt try the `streaming` option, but if you think is sufficient, we can pass this PR.

FrankD412 commented:
Another user was wondering on this also, so if you're okay with pushing this to the finish that would be welcomed. :)

danielafrimi commented:
/bot run

danielafrimi commented:
@rakib-hasan @FrankD412 Hi, rebased the branch. can we approve this PR (review has made before)?

tensorrt-cicd commented:
[PR_Github #9019](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9019/) [ run ] triggered by Bot

danielafrimi commented:
/bot run

tensorrt-cicd commented:
[PR_Github #9026](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9026/) [ run ] triggered by Bot

danielafrimi commented:
/bot kill

tensorrt-cicd commented:
[PR_Github #9028](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9028/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #9026](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9026/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #9031](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9031/) [ run ] triggered by Bot

FrankD412 commented on tensorrt_llm/bench/benchmark/throughput.py:
```diff
@@ -7,7 +7,6 @@
 import click
 from click_option_group import (MutuallyExclusiveOptionGroup, OptionGroup,
                                 optgroup)
-from huggingface_hub import snapshot_download
```
Ok I see the snapshot download was removed -- that has me worried because it's there due to a previous bug with the quant config not being present if the whole snapshot wasn't downloaded.

danielafrimi commented:
/bot run

tensorrt-cicd commented:
[PR_Github #9164](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9164/) [ run ] triggered by Bot

danielafrimi commented:
/bot run

tensorrt-cicd commented:
[PR_Github #9169](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9169/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #9686](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9686/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #9686](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9686/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #7121](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7121/) completed with status: 'FAILURE'

rakib-hasan commented on tensorrt_llm/bench/benchmark/throughput.py:
```diff
@@ -293,10 +293,11 @@ def throughput_command(
         # If we're dealing with a model name, perform a snapshot download to
         # make sure we have a local copy of the model.
-        if checkpoint_path is None:
+        if bench_env.checkpoint_path is None:
```
Re-iterating the comment from the other [PR](https://github.com/NVIDIA/TensorRT-LLM/commit/d0eb47d33a7195dfcb04e83cffd42518e6452bc6#r160748049).

@danielafrimi I ran into an issue with the previous code where the model is already downloaded and the code still tries to re-download since bench_env.checkpoint_path is None even though bench_env.model contains the local path.

Can you please elaborate on why do you think the current code is wrong?

tensorrt-cicd commented:
[PR_Github #9842](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9842/) [ run ] triggered by Bot

danielafrimi commented:
@FrankD412 will work this week on this PR. thanks

danielafrimi committed:
rebased

Signed-off-by: Daniel Afrimi <danielafrimi8@gmail.com>

danielafrimi commented:
/bot run

FrankD412 committed:
Merge branch 'main' into bench_latency

FrankD412 committed:
Update incorrect default in throughput.py

The default backend was set to None.

Not sure if this was a rebase error, but this needed to be fixed since None is not a valid option.

Signed-off-by: Frank <3429989+FrankD412@users.noreply.github.com>

FrankD412 commented:
/bot run

FrankD412 committed:
Merge branch 'main' into bench_latency

FrankD412 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #11840](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11840/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11841](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11841/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11840](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11840/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #11842](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11842/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8774](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8774/) completed with status: 'FAILURE'

FrankD412 committed:
Fix trtllm-bench test options.

This fixes a test issue where the model path was passed in for the model.

Signed-off-by: Frank <3429989+FrankD412@users.noreply.github.com>

FrankD412 committed:
Merge branch 'main' into bench_latency

FrankD412 committed:
Merge branch 'main' into bench_latency

FrankD412 committed:
Fix latency subcommand

Signed-off-by: Frank <3429989+FrankD412@users.noreply.github.com>

FrankD412 committed:
Merge branch 'main' into bench_latency

FrankD412 commented:
/bot run

FrankD412 committed:
Fix latency test.

Signed-off-by: Frank <3429989+FrankD412@users.noreply.github.com>

FrankD412 committed:
Fix throughput test.

Signed-off-by: Frank <3429989+FrankD412@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #11888](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11888/) [  ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #11886](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11886/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #11948](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11948/) [ run ] triggered by Bot

FrankD412 committed:
Merge branch 'main' into bench_latency

FrankD412 committed:
Merge branch 'main' into bench_latency

tensorrt-cicd commented:
[PR_Github #11964](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11964/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #11948](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/11948/) for commit `c3d44c0`

FrankD412 committed:
Merge branch 'main' into bench_latency

FrankD412 commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #11969](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11969/) [ reuse-pipeline ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11969](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11969/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #11948](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/11948/) for commit `b2f3c92`
